core/lib/zksync_core/src/consensus/testonly.rs

```diff
@@ -0,0 +1,291 @@
+            consensus: None,
+        });
+        self.actions_sender.push_actions(actions).await;
+        self.batch_sealed = true;
+    }
+
+    pub async fn push_random_blocks(&mut self, rng: &mut impl Rng, count: usize) {
+        for _ in 0..count {
+            // 20% chance to seal an L1 batch.
+            // seal_batch() also produces a (fictive) block.
+            if rng.gen_range(0..100) < 20 {
+                self.seal_batch().await;
+            } else {
+                self.push_block(rng.gen_range(3..8)).await;
+            }
+        }
+    }
+
+    // Wait for all pushed miniblocks to be produced.
+    pub async fn sync(&self, ctx: &ctx::Ctx, pool: &ConnectionPool) -> anyhow::Result<()> {
+        const POLL_INTERVAL: time::Duration = time::Duration::milliseconds(100);
+        loop {
+            let mut storage = pool.access_storage().await.context("access_storage()")?;
+            let head = storage
+                .blocks_dal()
+                .get_sealed_miniblock_number()
+                .await
+                .context("get_sealed_miniblock_number()")?;
+            if head.0 >= self.next_block.0 - 1 {
+                return Ok(());
+            }
+            ctx.sleep(POLL_INTERVAL).await?;
+        }
+    }
+
+    // Wait for all pushed miniblocks to have consensus certificate.
+    pub async fn sync_consensus(
+        &self,
+        ctx: &ctx::Ctx,
+        pool: &ConnectionPool,
+    ) -> anyhow::Result<()> {
+        const POLL_INTERVAL: time::Duration = time::Duration::milliseconds(100);
+        loop {
+            let mut storage = pool.access_storage().await.context("access_storage()")?;
+            if let Some(head) = storage
+                .blocks_dal()
+                .get_last_miniblock_number_with_consensus_fields()
+                .await
+                .context("get_last_miniblock_number_with_consensus_fields()")?
+            {
+                if head.0 >= self.next_block.0 - 1 {
+                    return Ok(());
+                }
+            }
+            ctx.sleep(POLL_INTERVAL).await?;
+        }
+    }
+
+    /// Validate consensus certificates for all expected miniblocks.
+    pub async fn validate_consensus(
+        &self,
+        ctx: &ctx::Ctx,
+        pool: &ConnectionPool,
+        genesis: validator::BlockNumber,
+        validators: &validator::ValidatorSet,
+    ) -> anyhow::Result<()> {
+        let mut storage = super::storage::storage(ctx, pool)
+            .await
+            .context("storage()")?;
+        for i in genesis.0..self.next_block.0 as u64 {
+            let block = storage
+                .fetch_block(ctx, validator::BlockNumber(i), self.operator_address)
+                .await?
+                .with_context(|| format!("missing block {i}"))?;
+            block.validate(validators, 1).unwrap();
+        }
+        Ok(())
+    }
+}
+
+async fn run_metadata_calculator(ctx: &ctx::Ctx, pool: ConnectionPool) -> anyhow::Result<()> {
```

Comment:
Nit: IMO, `run_mock_metadata_calculator` would be a better name; otherwise, one could get an idea that a real `MetadataCalculator` should be run here.

Reply:
done